### ISSUE TITLE
deadlock fair::(recursive_)timed_mutex::try_lock_{for,until}

### ISSUE THREAD

**Issue by yohhoy** [yohhoy/yamc#5]
commit aee4958549959e67457dba81b0944a11c641fcfe

The following tests cause deadlock in test/basic_test.
- TimedMutexTest/2.TryLockFor
- TimedMutexTest/2.TryLockUntil
- TimedMutexTest/3.TryLockFor
- TimedMutexTest/3.TryLockUntil

**Comment by yohhoy** [comment]
candidate patch:

```patch
diff --git a/include/fair_mutex.hpp b/include/fair_mutex.hpp
index f0b75f6..55cf2cc 100644
--- a/include/fair_mutex.hpp
+++ b/include/fair_mutex.hpp
@@ -160,15 +160,16 @@ class timed_mutex : private detail::mutex_base {
   bool do_try_lockwait(const std::chrono::time_point<Clock, Duration>& tp)
   {
     std::unique_lock<decltype(mtx_)> lk(mtx_);
-    std::size_t request = next_;
+    std::size_t request = next_++;
     while (request != curr_) {
       if (cv_.wait_until(lk, tp) == std::cv_status::timeout) {
         if (request == curr_)  // re-check predicate
           break;
+        --next_;
+        cv_.notify_all();
         return false;
       }
     }
-    ++next_;
     return true;
   }
```

**Comment by yohhoy** [comment]
This is an another draft patch to resolve this problem. In this algorithm, the order of mutex lock acquisition _may not_ be strict FIFO scheduling. `skip_` denotes timeouted `try_lock_*` request count (except timeouted thread was next candidate.), and it relax the condition of lock acquisition from "next candidate only" to "any thread in next + timeouted-count range".

```patch
diff --git a/include/fair_mutex.hpp b/include/fair_mutex.hpp
index f0b75f6..0eb845f 100644
--- a/include/fair_mutex.hpp
+++ b/include/fair_mutex.hpp
@@ -156,19 +156,31 @@ public:
 class timed_mutex : private detail::mutex_base {
   using base = detail::mutex_base;
 
+  std::size_t skip_ = 0;
+
   template <typename Clock, typename Duration>
   bool do_try_lockwait(const std::chrono::time_point<Clock, Duration>& tp)
   {
     std::unique_lock<decltype(mtx_)> lk(mtx_);
-    std::size_t request = next_;
-    while (request != curr_) {
+    std::size_t request = next_++;
+    while (!(curr_ <= request && request <= curr_ + skip_)) {
       if (cv_.wait_until(lk, tp) == std::cv_status::timeout) {
-        if (request == curr_)  // re-check predicate
+        if (curr_ <= request && request <= curr_ + skip_)  // re-check predicate
           break;
+        if (request == curr_ + 1) {
+          ++curr_;
+        } else {
+          ++skip_;
+        }
         return false;
       }
     }
-    ++next_;
+    assert(curr_ <= request && request <= curr_ + skip_);
+    if (0 < skip_ && next_ == curr_ + skip_ + 1) {
+      // there are no other waiting thread
+      skip_ = 0;
+      curr_ = next_ - 1;
+    }
     return true;
   }
```

**Comment by yohhoy** [comment]
Negatively resolved by commit 3e064f16e22fc99a65cf58a9719efa2619240813 that remove timed variant in `yamc::fair::*`.

I think that it's impossible to ensure perfect FIFO scheduling with ticked-based implementation. When mutex use waiting thread queue (Ex. `std::deque<std::thread::id>`), we can achieve stable FIFO lock order. But its run-time cost is not negligible overhead, and I feel that it's worthless.
